### PR TITLE
change(mac): store data in Library directory instead of Documents

### PR DESCRIPTION
**Pull request by sgschantz**
Store all keyman keyboards in `~/Library/Application Support/keyman.inputmethod.Keyman/Keyman-Keyboards`
instead of the old location `~/Documents/Keyman-Keyboards`

The new location is more appropriate because it is the standard location [designated by Apple](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/MacOSXDirectories/MacOSXDirectories.html).

Also, beginning with macOS Catalina in 2019, access to the /Documents directory by an application required the consent of the user. The new location will not require special permission.

This change involves 
1) moving all the keyboard data files to the new location and 
2) updating the settings stored in the UserDefaults for Keyman

The UserDefaults path settings need updating for the selected keyboard, the list of active keyboards and the list of keyboards with persisted options.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by mcdurdin** [comment]
This seems like it needs a bunch of user tests around install and upgrade scenarios.

**Comment by sgschantz** [comment]
> This seems like it needs a bunch of user tests around install and upgrade scenarios.

The user tests are going to be under #12144. (The first one is there.) That is the second phase of this change, and I wanted to avoid doing the user tests until both parts are in place because the data format is changing from the first to the second.

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I like this. I have some nits, mostly comments I think are out of sync with implementation in a few places, but one thing that looks like a bug

File: mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+ * This is in contrast with the lighter weight Settings which is stored using UserDefaults and handled by KMSettingsRepository.
+ */
+
+#import "KMDataRepository.h"
+#import "KMLogs.h"
+
+@interface KMDataRepository ()
+@property (readonly) NSURL *applicationSupportSubDirectory;   // '~/Library/Application Support'
+@property (readonly) NSURL *documentsSubDirectory;    // '~/Documents'
+@property (readonly) NSURL *obsoleteKeymanKeyboardsDirectory; // '~/Library/Documents/Keyman-Keyboards'
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:20]: The documentsSubDirectory is the user's Documents directory, identified using an API. The obsolete directory is the full path to the place we formerly stored data under the Documents directory.

Looking at the code, I can eliminate the property documentsSubDirectory as it is only used in finding the full obsolete data path. Would that help simplify this? It is a special location, so I thought it would be useful to save the reference, but we really never want to access it again after the migration.

I'll change the comments to further clarify.

[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:20]: After working on the comments a little. I think I need a better comment that explains the hierarchical relationship between these. We have three directories we should know of going forward and two obsolete ones.

File: mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+@synthesize keymanKeyboardsDirectory = _keymanKeyboardsDirectory;
+@synthesize obsoleteKeymanKeyboardsDirectory = _obsoleteKeymanKeyboardsDirectory;
+@synthesize keymanDataDirectory = _keymanDataDirectory;
+
+NSString *const kKeyboardsDirectoryName = @"Keyman-Keyboards";
+/* 
+ The name of the subdirectory within '~/Library/Application Support'.
+ The convention is to use bundle identifier.
+ (Using the subsystem id, "com.keyman.app", is a poor choice because the API
+ createDirectoryAtPath sees the .app extension and creates an application file.
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:36]: Yes, I'll fix the comments which are incorrect. I prefer `com.keyman.app` as an app ID but since our app ID is already set to `keyman.inputmethod.Keyman` and `com.keyman.app` looks to the file manager like a filename of type `.app`, I plant to stick with `keyman.inputmethod.Keyman` as the name of our app's data directory.

File: mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:20]: Yes, sounds good, thanks

File: mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      os_log_info([KMLogs dataLog], "Application Support subdirectory: '%{public}@'", applicationSupportUrl.path);
+      _applicationSupportSubDirectory = applicationSupportUrl;
+    }
+  }
+  return _applicationSupportSubDirectory;
+}
+
+- (NSURL *)keymanDataDirectory {
+  if (_keymanDataDirectory == nil) {
+    NSURL *keymanDataUrl = [self.applicationSupportSubDirectory URLByAppendingPathComponent:kKeymanSubdirectoryName  isDirectory: TRUE];
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:86]: I cannot think of anything intelligent to do other than log an error if the OS does not return it to me. This is the place that we are supposed to create our app subdirectory for our user data, so if it doesn't exist, then we don't have a fallback location that I am aware of. In my 'Application Support' directory, there are 91 subdirectories created by other software, so it would be a major failure if it weren't there.

File: mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+- (NSURL *)keymanKeyboardsDirectory {
+  if (_keymanKeyboardsDirectory == nil) {
+    NSURL *keyboardsUrl = [self.keymanDataDirectory URLByAppendingPathComponent:kKeyboardsDirectoryName  isDirectory: TRUE];
+    _keymanKeyboardsDirectory = keyboardsUrl;
+  }
+  return _keymanKeyboardsDirectory;
+}
+
+/**
+ * Creates Keyman data directory if it do not exist yet. This is the main data subdirectory: keyman.inputmethod.Keyman
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:101]: The header was wrong. 😒

File: mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  return exists;
+}
+
+/**
+ * Migrate the keyboards data from the old location in '/Documents' to the new location '/Application Support/keyman.inputmethod.Keyman/'
+ */
+- (BOOL)migrateData {
+  BOOL didMoveData = NO;
+  NSFileManager *fileManager = [NSFileManager defaultManager];
+  BOOL dataExistsInOldLocation = [self keyboardsExistInObsoleteDirectory];
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m:166]: Yes, I added a comment for this and another function called only by migrateData that would also trigger the access request. 

This would only happen if the Keyman was configured to store data in ~/Documents (so that the settings in UserDefaults were updated with the location) but then they revoked those privileges just before installing the new version of Keyman that would attempt the migration.

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * KMSettingsRepository.h
+ * Keyman
+ *
+ * Created by Shawn Schantz on 2024-07-29.
+ *
+ */
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h:9]: Ugh, I thought I was following a pattern that I copied from the standards a while back.  I thought we even had a discussion about the date format. 

I think that author and date can be very helpful to understand the context of the code at a glance without jumping into git. We can talk separately about this, but Is there a strong reason to exclude those?

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+ * 2) the value  for KMStoreKeyboardsInLibraryKey is not set to true
+ */
+- (BOOL)dataMigrationNeeded {
+  BOOL keymanSettingsExist = [self settingsExist];
+  os_log([KMLogs dataLog], "  keyman settings exist: %{public}@", keymanSettingsExist ? @"YES" : @"NO" );
+  
+  BOOL dataInLibrary = [self dataStoredInLibraryDirectory];
+  os_log([KMLogs dataLog], "  data stored in Library: %{public}@", dataInLibrary ? @"YES" : @"NO" );
+  
+  return !(keymanSettingsExist && dataInLibrary);
```
[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m:81]: Actually, this works, but it just looks wrong because of a poor method name. I'll rename it.

That method is only checking the settings, not attempting to access the actual data. Everything in the KMSettingsRepository class only accesses UserDefaults and not the file system.

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h:9]: > I think that author and date can be very helpful to understand the context of the code at a glance without jumping into git. We can talk separately about this, but Is there a strong reason to exclude those?

The primary reasons _not_ to do it are:

1. avoid multiple "sources of truth" -- the initial creator of the file may well not be the most relevant, nor the create date
2. avoid busy work

I think we can be pretty confident that git will get it right. In most good code editors, we can see git blame line-by-line, which nicely eliminates the burden of seeing the git data. I would encourage going to git for this kind of metadata -- commit history is really helpful (probably more helpful than original author + date IMO).

The only real downside I can see is if you want to see the original commit: if the header changes over time, then the git blame will show most recent commit against it by default, not the original, and it takes a little more digging to get back to the first commit.

[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h:9]: I don't think any of these things override the practical value of this tiny piece of metadata: who created this thing and when. It is written once in the life of the file, and only ever updated with maybe a second author if there is a major redesign within the same code. 

In most cases, it will remain as one name and one date but can reveal a lot on its own without scanning the commit history. If it doesn't seem helpful, it can easily be ignored along with the copyright statement. In my experience, the initial author is heavily involved in the design of the code (often the sole person responsible), and a lot of the original design persists even after many edits and attempts to refactor away from the original design.

For the Mac code at least, it looks like the first commit was not the author at all but just the guy who first made use of git. So the real author would be anonymous if he hadn't included his name in the header.

In short, I don't see this as redundant information but something with long-term practical value and maintenance cost that is approximately zero.

The file name and project name does seem like redundant busywork in the header, so I agree that we should remove those. (I've changed the class name many times and forgotten to update the header.)

[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h:9]: OK I am not overly bothered either way. Personally I feel like it is busy work which replicates what we _now_ have in git. Yes, back in the day we didn't use git but that's a slippery argument because then we end up putting all commit history back into source files because what if we change vcs in the future and cannot transfer commit history and well yuck.

See for example https://github.com/keymanapp/keyman/blob/d8986be12fe0fb2078b93fa4650e2a7fe06f16d1/windows/src/engine/keyman32/keyman32.cpp for what things looked like before we relied on git history. If you scan through the file you'll also see `// I1234` comments which were how we tagged changes. It worked but it was increasingly noisy so it was really good to move that change history into vcs metadata.

tl;dr: If you want to keep it in, OK, and then please also update the code style wiki to say that we can optionally include the `Created by <author> on yyyy-mm-dd` line.

[the remark this answers is not in the document]
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.h:9]: Thanks, I find that one extra line to be helpful. I've updated the wiki.

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m [hunk 8]
```diff
@@ -0,0 +1,211 @@
+/*
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * KMSettingsRepository.h
+ * Keyman
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m:5]:
```suggestion
```

Case in point - this says `KMSettingsRepository.h` but it is in a `.m` :grin:

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+- (BOOL)settingsExist {
+  return [[NSUserDefaults standardUserDefaults] objectForKey:kSelectedKeyboardKey] != nil;
+}
+
+/**
+ * For the first numbered version of the data model, the app stores the keyboards under the ~/Library directory
+ * For versions before version 1, the keyboards were stored under the ~/Documents directory.
+ */
+- (BOOL)dataModelWithKeyboardsInLibrary {
+  // NSUserDefaults returns zero if the key does not exist
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m:64]:
```suggestion
  // [NSUserDefaults integerForKey] returns zero if the key does not exist
```

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      os_log([KMLogs dataLog], "converted selected keyboard setting from '%{public}@' to '%{public}@'", selectedKeyboardPath, newPathString);
+    }
+  }
+}
+
+/**
+ * Convert the path of the keyboard designating the Documents folder to its new location
+ * in the Application Support folder
+ */
+
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m:145]:
```suggestion
```

**Review by mcdurdin — APPROVED** [review]
LGTM, changes look good, formatting and comment nits only now!

File: mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m (hunk 9 above)
Comment (sgschantz) [reply mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m:64]: Created issue #12294 to document the need for better error handling in case the data directory cannot be located or is unreadable. The current code does not gracefully handle a nil returned from the OS when trying to access '~/Library/Application Support'

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.102-alpha](https://keyman.com/downloads/releases/alpha/18.0.102)
